File: packages/dotnet/Iscc.Lib.Tests/UnicodeBoundaryTests.cs
// Unicode 16.0.0 boundary conformance tests for the .NET binding.
//
// Runs the canonical boundary fixture (crates/iscc-lib/tests/unicode_boundary.json,
// linked into the test output directory by the csproj) against TextClean and
// TextCollapse, gating the declared-Unicode-version freeze rule: code points
// unassigned in Unicode 16.0.0 are mapped to the noncharacter sentinel U+FFFF
// before normalization and removed by the unchanged category-C filter, while
// assigned code points flow through the regular pipeline. The sequence vectors
// additionally pin the sentinel map against the superseded delete-filter design.

using System.Text.Json;
using Iscc.Lib;
using Xunit;

namespace Iscc.Lib.Tests;

/// <summary>Validates TextClean and TextCollapse against the canonical Unicode boundary vectors.</summary>
public class UnicodeBoundaryTests
{
    /// <summary>Cached parsed unicode_boundary.json for all test methods.</summary>
    private static readonly Lazy<JsonElement> Boundary = new(LoadBoundary);

    /// <summary>Load and parse testdata/unicode_boundary.json from the test output directory.</summary>
    private static JsonElement LoadBoundary()
    {
        string path = Path.Combine(AppContext.BaseDirectory, "testdata", "unicode_boundary.json");
        string json = File.ReadAllText(path);
        return JsonDocument.Parse(json).RootElement;
    }

    /// <summary>Guard against a truncated or renamed fixture silently defining zero vector tests.</summary>
    [Fact]
    public void BoundaryFixtureMetadata()
    {
        JsonElement metadata = Boundary.Value.GetProperty("_metadata");
        Assert.Equal("16.0.0", metadata.GetProperty("unicode_data_version").GetString());
        Assert.Equal(7, Boundary.Value.GetProperty("text_clean").EnumerateObject().Count());
        Assert.Equal(5, Boundary.Value.GetProperty("text_collapse").EnumerateObject().Count());
    }

    // ── text_clean ───────────────────────────────────────────────────────────

    /// <summary>Yield boundary cases for text_clean (7 vectors).</summary>
    public static IEnumerable<object[]> TextCleanVectors()
    {
        foreach (JsonProperty kv in Boundary.Value.GetProperty("text_clean").EnumerateObject())
            yield return [kv.Name, kv.Value];
    }

    [Theory]
    [MemberData(nameof(TextCleanVectors))]
    public void TextCleanBoundary(string _, JsonElement tc)
    {
        string input = tc.GetProperty("inputs")[0].GetString()!;
        string expected = tc.GetProperty("outputs").GetProperty("result").GetString()!;
        Assert.Equal(expected, IsccLib.TextClean(input));
    }

    // ── text_collapse ────────────────────────────────────────────────────────

    /// <summary>Yield boundary cases for text_collapse (5 vectors).</summary>
    public static IEnumerable<object[]> TextCollapseVectors()
    {
        foreach (JsonProperty kv in Boundary.Value.GetProperty("text_collapse").EnumerateObject())
            yield return [kv.Name, kv.Value];
    }

    [Theory]
    [MemberData(nameof(TextCollapseVectors))]
    public void TextCollapseBoundary(string _, JsonElement tc)
    {
        string input = tc.GetProperty("inputs")[0].GetString()!;
        string expected = tc.GetProperty("outputs").GetProperty("result").GetString()!;
        Assert.Equal(expected, IsccLib.TextCollapse(input));
    }
}

File: packages/swift/Tests/IsccLibTests/UnicodeBoundaryTests.swift
/// Unicode 16.0.0 boundary vector tests for textClean / textCollapse.
/// Gates the declared-Unicode-version freeze rule (code points unassigned in Unicode 16.0.0
/// map to the U+FFFF sentinel before normalization) against the vendored
/// unicode_boundary.json fixture — 7 text_clean + 5 text_collapse vectors.

import Foundation
import XCTest
@testable import IsccLib

final class UnicodeBoundaryTests: XCTestCase {

    /// Declared Unicode data version pinned by the fixture metadata.
    static let expectedUnicodeVersion = "16.0.0"

    /// Expected number of text_clean boundary vectors.
    static let expectedTextCleanCount = 7

    /// Expected number of text_collapse boundary vectors.
    static let expectedTextCollapseCount = 5

    /// Cached parsed unicode_boundary.json for all test methods.
    static var fixture: [String: Any] = {
        let url = Bundle.module.url(forResource: "unicode_boundary", withExtension: "json")!
        let data = try! Data(contentsOf: url)
        return try! JSONSerialization.jsonObject(with: data) as! [String: Any]
    }()

    /// Load boundary vectors for a given fixture section, sorted by case name.
    static func vectors(for section: String) -> [(String, [String: Any])] {
        guard let group = fixture[section] as? [String: Any] else { return [] }
        return group.sorted { $0.key < $1.key }.map { ($0.key, $0.value as! [String: Any]) }
    }

    /// Unicode scalar values of a string, for exact code-point comparison.
    ///
    /// Swift's `String ==` (and therefore a bare `XCTAssertEqual` on strings) folds
    /// canonical equivalence: "e" + U+0301 == U+00E9 is true. That would make the
    /// sequence vectors vacuous — they exist precisely to distinguish the U+FFFF
    /// sentinel map from a delete filter, and both designs produce canonically
    /// equivalent strings for three of the four sequences. Always compare scalar
    /// arrays here, never strings directly.
    static func scalars(_ text: String) -> [UInt32] {
        return text.unicodeScalars.map { $0.value }
    }

    /// Run every vector in a section through a transform and assert scalar-exact output.
    static func assertVectors(
        section: String,
        transform: (String) -> String,
        file: StaticString = #filePath,
        line: UInt = #line
    ) {
        for (name, tc) in vectors(for: section) {
            let inputs = tc["inputs"] as! [Any]
            let outputs = tc["outputs"] as! [String: Any]
            let input = inputs[0] as! String
            let expected = outputs["result"] as! String
            let actual = transform(input)
            XCTAssertEqual(
                scalars(actual), scalars(expected),
                "Failed vector: \(section)/\(name)", file: file, line: line)
        }
    }

    /// Verify the fixture pins the declared Unicode version and both vector counts.
    func testFixtureMetadata() {
        let metadata = Self.fixture["_metadata"] as! [String: Any]
        XCTAssertEqual(
            metadata["unicode_data_version"] as! String, Self.expectedUnicodeVersion,
            "Fixture Unicode data version drifted")
        XCTAssertEqual(
            Self.vectors(for: "text_clean").count, Self.expectedTextCleanCount,
            "Unexpected text_clean vector count")
        XCTAssertEqual(
            Self.vectors(for: "text_collapse").count, Self.expectedTextCollapseCount,
            "Unexpected text_collapse vector count")
    }

    /// All 7 text_clean boundary vectors must pass with scalar-exact output.
    func testTextCleanBoundaryVectors() {
        Self.assertVectors(section: "text_clean", transform: { textClean(text: $0) })
    }

    /// All 5 text_collapse boundary vectors must pass with scalar-exact output.
    func testTextCollapseBoundaryVectors() {
        Self.assertVectors(section: "text_collapse", transform: { textCollapse(text: $0) })
    }
}
